### PR TITLE
feat(B-0276): structure fingerprint library for codebase shape recognition

### PR DESCRIPTION
**Pull request by AceHack**
## Summary

- Adds `src/Core/StructureFingerprint.fs` with 10 canonical architecture shapes (operator-algebra, state-machine, pipeline, pub-sub, tree-hierarchy, layered-dag, star-topology, ring, bipartite, unknown)
- Structural signal extraction from `Graph<'N>` (node/edge counts, degree distribution, cycle detection, layer counting, bipartiteness)
- Heuristic classifier with confidence scores, shape similarity comparison, and kebab-case labels
- 16 xUnit tests covering all recognized shapes, signal extraction, similarity semantics, and edge cases

## Backlog

Closes B-0276 (child of B-0240 structure recognizer). First graduation: types + classification. Future: spectral/topological fingerprinting (B-0240 Layer 1-3).

## Test plan

- [x] `dotnet build -c Release` — 0 warnings, 0 errors
- [x] 16 tests pass: pipeline, ring, star, tree, layered DAG, state machine, pub/sub, empty graph, signal extraction, similarity, shape labels
- [ ] CI confirms on merge

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/Core/StructureFingerprint.fs [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+               && signals.RootCount = 1
+               && e = n - 1
+               && signals.MaxOutDegree > 1L
+            then 0.9
+            else 0.0
+
+        // State machine: cycles present, moderate connectivity
+        let stateScore =
+            if signals.HasCycles && n >= 2 && e > n then 0.7
+            elif signals.HasCycles && n >= 2 then 0.5
```
[the remark this answers is not in the document]
Comment (AceHack) [reply src/Core/StructureFingerprint.fs:222]: Fixed in 9fd4353. OperatorAlgebra score raised to 0.75 (bipartite-cyclic path) / 0.70 (dense-cyclic path), and reordered before StateMachine in the candidates list. Added a test with a dense cyclic bipartite graph that now correctly classifies as OperatorAlgebra.

File: src/Core/StructureFingerprint.fs [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    /// Compute topological layers via BFS from roots (nodes with
+    /// in-degree 0). Returns the number of layers. If the graph
+    /// has cycles or no roots, returns 0.
+    let private layerCount (g: Graph<'N>) : int =
+        let allNodes = Graph.nodes g
+        let roots =
+            allNodes
+            |> Set.filter (fun n -> Graph.inNeighbors n g |> List.isEmpty)
+        if roots.IsEmpty then 0
+        else
```
[the remark this answers is not in the document]
Comment (AceHack) [reply src/Core/StructureFingerprint.fs:97]: Fixed in 9fd4353. `layerCount` now calls `hasCycle g` first and returns 0 for cyclic graphs, matching the documented contract.

File: src/Core/StructureFingerprint.fs [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+                        elif visiting.Contains node then
+                            found <- true
+                        else
+                            visiting <- visiting.Add node
+                            stack.Push(node, true)
+                            for (neighbor, _w) in Graph.outNeighbors node g do
+                                if visiting.Contains neighbor then
+                                    found <- true
+                                elif not (visited.Contains neighbor) then
+                                    stack.Push(neighbor, false)
```
[the remark this answers is not in the document]
Comment (AceHack) [reply src/Core/StructureFingerprint.fs:83]: Investigated — the original algorithm is actually correct for this case. The post-order marking (mark `visiting` on pop, not push) prevents the false positive: in a diamond DAG 0→1, 0→2, 2→1, node 1 gets pushed twice but only enters `visiting` when first popped; the second pop finds it in `visited` and skips. Added a diamond-DAG test to confirm this invariant holds.

File: src/Core/StructureFingerprint.fs [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+                for (neighbor, _w) in Graph.outNeighbors node g do
+                    if not (depth.ContainsKey neighbor) then
+                        depth.[neighbor] <- d + 1
+                        queue.Enqueue neighbor
+            if depth.Count = 0 then 0
+            else
+                let mutable maxD = 0
+                for kvp in depth do
+                    if kvp.Value > maxD then maxD <- kvp.Value
+                maxD + 1
```
[the remark this answers is not in the document]
Comment (AceHack) [reply src/Core/StructureFingerprint.fs:115]: Fixed in 9fd4353. `layerCount` now returns 0 when `hasCycle g` is true, matching the doc comment contract.

File: tests/Tests.FSharp/Algebra/StructureFingerprint.Tests.fs [hunk 5]
```diff
@@ -0,0 +1,179 @@
+module Zeta.Tests.Algebra.StructureFingerprintTests
+
```
[the remark this answers is not in the document]
Comment (AceHack) [reply tests/Tests.FSharp/Algebra/StructureFingerprint.Tests.fs:3]: Fixed in 9fd4353. Added `#nowarn "0893"` at line 2.

File: src/Core/StructureFingerprint.fs [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+        let candidates =
+            [ (StarTopology, starScore)
+              (Pipeline, pipelineScore)
+              (Ring, ringScore)
+              (TreeHierarchy, treeScore)
+              (LayeredDag, layeredScore)
+              (PubSub, pubsubScore)
+              (StateMachine, stateScore)
+              (OperatorAlgebra, algebraScore)
+              (Bipartite, bipartiteScore) ]
```
[the remark this answers is not in the document]
Comment (AceHack) [reply src/Core/StructureFingerprint.fs:268]: Fixed in 9fd4353. Added three new tests: `OperatorAlgebra` (dense cyclic bipartite graph), `Bipartite` (many-source few-sink acyclic bipartite), and a diamond-DAG cycle-detection correctness test.
